source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,159 @@
+#include "source/opt/function.h"
+#include "source/opt/module.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/opt/instruction.h"
+#include "source/fuzz/data_descriptor.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "test/fuzz/fuzz_test_util.h"
+#include <algorithm>
+#include "source/fuzz/fact_manager/fact_manager.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    protobufs::TransformationWrapVectorSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(uint32_t instruction_id, uint32_t vec_id1, uint32_t vec_id2,
+                                                                 uint32_t vec_id3,  uint32_t vec_type_id, uint32_t pos,
+                                                                 std::vector<uint32_t>& vec1_elements, std::vector<uint32_t>& vec2_elements) {
+    message_.set_instruction_id(instruction_id);
+    message_.set_vec_id1(vec_id1);
+    message_.set_vec_id2(vec_id2);
+    message_.set_vec_id3(vec_id3);
+    message_.set_vec_type_id(vec_type_id);
+    message_.set_scalar_position(pos);
+    for(auto id : vec1_elements) {
+      message_.add_vec1_elements(id);
+    }
+    for(auto id : vec2_elements) {
+      message_.add_vec2_elements(id);
+    }
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+      opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+    auto valid_arithmetic_types = std::unordered_set<SpvOp> {SpvOpIAdd, SpvOpISub, SpvOpIMul, SpvOpFAdd, SpvOpFSub, SpvOpFMul};
+    auto instruction = ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+    auto vector_type = ir_context->get_def_use_mgr()->GetDef(message_.vec_type_id());
+
+    // |instruction_id| must refer to an existing instruction.
+    if(instruction == nullptr) return false;
+
+    // |vec_type_id| must refer to a valid type instruction.
+    if(vector_type == nullptr) return false;
+
+    // |instruction_id| must be a valid arithmetic type.
+    assert(valid_arithmetic_types.count(instruction->opcode()) && "The instruction must be of a valid arithmetic type.");
+
+    // |vector_type_id| must correspond to a valid vector type.
+    assert((vector_type->opcode() == SpvOpTypeVector) && "Vector type id must correspond to a valid vector type.");
+
+    // |vec_id1|, |vec_id2| and |vec_id3| must be fresh.
+    if(!fuzzerutil::IsFreshId(ir_context, message_.vec_id1()) || !fuzzerutil::IsFreshId(ir_context, message_.vec_id2())
+        || !fuzzerutil::IsFreshId(ir_context, message_.vec_id3())) return false;
+
+    // |vec_id1|, |vec_id2| and |vec_id3| should have disparate ids.
+    if(message_.vec_id1() == message_.vec_id2() || message_.vec_id2() == message_.vec_id3() || message_.vec_id1() == message_.vec_id3()) return false;
+
+    // |pos| needs to be a non-negative integer less than the vector length.
```

Comment:
Should be |scalar_position|. I designed it so that the vector should have a zero constant id in that position and later we want the ids of the original arithmetic operation to be filled in the position.